packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,318 @@
+import { z, ZodType } from 'zod';
+import {
+  CallToolRequest,
+  CallToolResult,
+  CallToolResultSchema,
+  CompatibilityCallToolResult,
+  CompatibilityCallToolResultSchema,
+  Implementation,
+  InitializeResultSchema,
+  JSONRPCError,
+  JSONRPCNotification,
+  JSONRPCRequest,
+  JSONRPCResponse,
+  LATEST_PROTOCOL_VERSION,
+  ListToolsRequest,
+  ListToolsResult,
+  ListToolsResultSchema,
+  Notification,
+  Request,
+  RequestOptions,
+  ServerCapabilities,
+  SUPPORTED_PROTOCOL_VERSIONS,
+  TimeoutInfo,
+  Transport,
+} from './types';
+import { AISDKError } from '@ai-sdk/provider';
+
+/**
+ * A lightweight MCP Client implementation,
+ * primarily for tool conversion between MCP<>AI SDK.
+ *
+ * It is a custom implementation of the MCP Client (derived from Protocol) class.
+ *
+ * Not supported:
+ * - Client options (e.g. sampling, roots) as they are not needed for tool conversion
+ * - Accepting notifications
+ */
+export class SimpleMcpClient {
+  private _transport: Transport | undefined;
+  private _requestMessageId = 0;
+  private _responseHandlers: Map<
+    number,
+    (response: JSONRPCResponse | Error) => void
+  > = new Map();
+  private connectionTimeoutMs: number;
+  private requestTimeoutMs: number;
+  private _timeoutInfo: Map<number, TimeoutInfo> = new Map();
+  private _serverCapabilities?: ServerCapabilities;
+
+  constructor(private _clientInfo: Implementation) {
+    this.connectionTimeoutMs = _clientInfo.connectionTimeoutMs || 6000;
+    this.requestTimeoutMs = _clientInfo.requestTimeoutMs || 3000;
+  }
+
+  async connect(transport: Transport): Promise<void> {
```

Comment:
merge into constructor?

Reply:
brilliant!

Reply:
ope, silly me: we can't have an async constructor - lmk if i'm missing something!